xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,470 @@
+        if (newWeight > 0) {
+          avgWeight += newWeight;
+          weightedChannelCount++;
+        }
+      }
+      if (weightedChannelCount < 2) {
+        rrMode = true;
+        return;
+      }
+      avgWeight /= 1.0 * weightedChannelCount;
+      for (int i = 0; i < list.size(); i++) {
+        WrrSubchannel subchannel = (WrrSubchannel) list.get(i);
+        double newWeight = subchannel.getWeight();
+        scheduler.add(i, newWeight > 0 ? newWeight : avgWeight);
+      }
+      schedulerRef.set(scheduler);
+      rrMode = false;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(WeightedRoundRobinPicker.class)
+          .add("list", list).add("rrMode", rrMode).toString();
+    }
+
+    @VisibleForTesting
+    List<Subchannel> getList() {
+      return list;
+    }
+
+    @Override
+    public boolean isEquivalentTo(RoundRobinPicker picker) {
+      if (!(picker instanceof WeightedRoundRobinPicker)) {
+        return false;
+      }
+      WeightedRoundRobinPicker other = (WeightedRoundRobinPicker) picker;
+      // the lists cannot contain duplicate subchannels
+      return other == this
+          || (list.size() == other.list.size() && new HashSet<>(list).containsAll(other.list));
+    }
+  }
+
+  /**
+   * The earliest deadline first implementation in which each object is
+   * chosen deterministically and periodically with frequency proportional to its weight.
+   *
+   * <p>Specifically, each object added to chooser is given a deadline equal to the multiplicative
+   * inverse of its weight. The place of each object in its deadline is tracked, and each call to
+   * choose returns the object with the least remaining time in its deadline (1/weight).
+   * (Ties are broken by the order in which the children were added to the chooser.)
+   * For example, if items A and B are added with weights 0.5 and 0.2, successive chooses return:
+   *
+   * <ul>
+   *   <li>In the first call, the deadlines are A=2 (1/0.5) and B=5 (1/0.2), so A is returned.
+   *   The deadline of A is updated to 4.
+   *   <li>Next, the remaining deadlines are A=4 and B=5, so A is returned. The deadline of A (2) is
+   *       updated to A=6.
+   *   <li>Remaining deadlines are A=6 and B=5, so B is returned. The deadline of B is updated with
+   *       with B=10.
+   *   <li>Remaining deadlines are A=6 and B=10, so A is returned. The deadline of A is updated with
+   *        A=8.
+   *   <li>Remaining deadlines are A=8 and B=10, so A is returned. The deadline of A is updated with
+   *       A=10.
+   *   <li>Remaining deadlines are A=10 and B=10, so A is returned. The deadline of A is updated
+   *      with A=12.
+   *   <li>Remaining deadlines are A=12 and B=10, so B is returned. The deadline of B is updated
+   *      with B=15.
+   *   <li>etc.
+   * </ul>
+   *
+   * <p>In short: the entry with the highest weight is preferred. In case of ties, the object that
```

Comment:
This disagrees with `o1.index - o2.index` in the comparitor. Personally, the language here sounds good and I'd get rid of the index subtraction, but it's up to you except that one of the two need changing.